src/components/table-of-contents/table-of-contents.js

```diff
@@ -1,38 +1,50 @@
 export default class TableOfContents {
     constructor(component) {
         this.component = component;
-        this.sections = [...this.component.querySelectorAll('section[id]')];
-        this.refreshIntervalId = setInterval(() => this.setCurrent(), 100);
-        this.setCurrent();
-    }
+        this.sections = [...this.component.querySelectorAll('section[id]')].filter((section) => section.id);
+        this.tocLinks = {};
+        this.activeSection = null;
 
-    setCurrent() {
-        let activeSection = this.sections[0];
-        for (let section of this.sections) {
-            const top = section.getBoundingClientRect().top;
-            if (top > 100) {
-                break;
+        this.sections.forEach((section) => {
+            const link = this.component.querySelector(`.ons-list__link[href="#${section.id}"]`);
+            if (link) {
+                this.tocLinks[section.id] = link;
             }
+        });
 
-            activeSection = section;
+        this.observer = new IntersectionObserver(this.handleIntersect.bind(this), {
+            rootMargin: '0px 0px -100% 0px', // trigger when top of section is at the top of viewport
+        });
 
-            if (top >= 0 && top <= 100) {
-                break;
-            }
-        }
+        this.sections.forEach((section) => this.observer.observe(section));
+    }
 
-        if (activeSection === this.activeSection) {
-            return;
-        }
+    handleIntersect(entries) {
+        // Find the topmost visible section
+        const visibleSections = entries
+            .filter((entry) => entry.isIntersecting)
+            .sort((a, b) => a.target.getBoundingClientRect().top - b.target.getBoundingClientRect().top);
+
+        if (visibleSections.length === 0) return;
+
+        const newActive = visibleSections[0].target;
+
+        if (newActive === this.activeSection) return;
+
+        this.activeSection = newActive;
+        this.updateTocLinks();
+    }
+
+    updateTocLinks() {
+        for (const section of this.sections) {
+            const link = this.tocLinks[section.id];
 
-        this.activeSection = activeSection;
+            if (!link) continue;
```

Comment:
This and the lines below can be simplified to:
```js
link?.classList.toggle('ons-table-of-contents__link-active', section === this.activeSection);
```

Reply:
👌

Reply:
Updated with a comment just for clarity, thanks @MaciekBaron